internal/log_analysis/alerts_api/table/list.go

```diff
@@ -19,93 +19,124 @@ package table
 
-	// queries require and = condition on primary key
-	keyCondition := expression.Key(ddbKey).Equal(expression.Value(&ddbValue))
-
-	queryExpression, err := expression.NewBuilder().
-		WithKeyCondition(keyCondition).
-		Build()
+	// Get the sort direction
+	direction := table.isAscendingOrder(input)
 
+	// Construct a query expression
+	queryExpression, err := builder.Build()
 	if err != nil {
 		return nil, nil, errors.Wrap(err, "failed to build expression")
 	}
 
+	// Limit the returned results to the specified page size or max default
 	var queryResultsLimit *int64
-	if pageSize != nil {
-		queryResultsLimit = aws.Int64(int64(*pageSize))
+	if input.PageSize != nil {
+		queryResultsLimit = aws.Int64(int64(*input.PageSize))
+	} else {
+		queryResultsLimit = aws.Int64(int64(25))
 	}
 
+	// Optionally continue the query from the "primary key of the item where the [previous] operation stopped"
 	var queryExclusiveStartKey map[string]*dynamodb.AttributeValue
-	if exclusiveStartKey != nil {
+	if input.ExclusiveStartKey != nil {
 		queryExclusiveStartKey = make(map[string]*dynamodb.AttributeValue)
-		err = jsoniter.UnmarshalFromString(*exclusiveStartKey, &queryExclusiveStartKey)
+		err = jsoniter.UnmarshalFromString(*input.ExclusiveStartKey, &queryExclusiveStartKey)
 		if err != nil {
 			return nil, nil, errors.Wrap(err, "failed to Unmarshal ExclusiveStartKey")
 		}
 	}
 
+	// Construct the full query
 	var queryInput = &dynamodb.QueryInput{
 		TableName:                 &table.AlertsTableName,
-		ScanIndexForward:          aws.Bool(false),
+		ScanIndexForward:          aws.Bool(direction),
 		ExpressionAttributeNames:  queryExpression.Names(),
 		ExpressionAttributeValues: queryExpression.Values(),
+		FilterExpression:          queryExpression.Filter(),
 		KeyConditionExpression:    queryExpression.KeyCondition(),
 		ExclusiveStartKey:         queryExclusiveStartKey,
-		IndexName:                 aws.String(index),
-		Limit:                     queryResultsLimit,
+		IndexName:                 index,
+		Limit:                     aws.Int64(*queryResultsLimit * 4), //optimization accounting for filtering
 	}
 
-	queryOutput, err := table.Client.Query(queryInput)
-	if err != nil {
-		// this deserves detailed logging for debugging
-		zap.L().Error("Query()", zap.Error(err), zap.Any("input", queryInput), zap.Any("startKey", queryExclusiveStartKey))
-		return nil, nil, errors.Wrapf(err, "QueryInput() failed for %s,%s", ddbKey, ddbValue)
-	}
+	// Define placeholder for the last key that we will send back to the frontend
+	var lastKey DynamoItem
+	// Continuously query until we have enough results for the requested page size
+	err = table.Client.QueryPages(queryInput, func(page *dynamodb.QueryOutput, isLast bool) bool {
+		for _, item := range page.Items {
+			// Define temp container
+			var alert *AlertItem
+			// Unmarshal each item to an alert
+			err = dynamodbattribute.UnmarshalMap(item, &alert)
+			if err != nil {
+				return false
+			}
```

Comment:
No need for a double-pointer, and you also need to handle that error condition instead of just ignoring it. One way to do that is to define a `var errMarshal error` outside the function closure (e.g. line 93). Then:

```suggestion
			var alert AlertItem
			// Unmarshal each item to an alert
			if err = dynamodbattribute.UnmarshalMap(item, &alert); err != nil {
				// Something is wrong with this Dynamo item
				marshalErr = err
				return false
			}
```

The way you have it now, the `err` variable would be overwritten with `nil` on line 94 when `QueryPages` returns successfully, even if there were an error inside the function. Error handling can sometimes get awkward inside a function closure.

Then you would check if `err != nil` or `marshalErr != nil`

Reply:
Good catch. I've now captured the error outside the closure.